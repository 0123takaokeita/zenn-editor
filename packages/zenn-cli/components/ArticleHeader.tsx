import React from 'react';
import { Article } from '@types';
import { ContentWrapper } from '@components/ContentWrapper';
import { ErrorRow } from '@components/ErrorRow';
import { getArticleErrors } from '@utils/validator';

type Props = { article: Article };

export const ArticleHeader: React.FC<Props> = ({ article }) => {
  const errorMessages = getArticleErrors(article);
  const errorCount = errorMessages?.length;
  return (
    <header className="content-header">
      <ContentWrapper>
        <h1 className="content-header__title">{article.title || 'No Title'}</h1>

        <div className="content-header__row">
          <span className="content-header__row-title">slug</span>
          <span className="content-header__row-result">{article.slug}</span>
        </div>

        <div className="content-header__row">
          <span className="content-header__row-title">emoji</span>
          <span className="content-header__row-result">
            {article.emoji ? (
              <span style={{ fontSize: '20px' }}>{article.emoji}</span>
            ) : (
              '指定が必要です'
            )}
          </span>
        </div>

        <div className="content-header__row">
          <span className="content-header__row-title">topics</span>
          <span className="content-header__row-result">
            {Array.isArray(article.topics) && article.topics.length
              ? article.topics.map((t, i) => (
                  <span className="content-header__topic" key={`at${i}`}>
                    {t}
                  </span>
                ))
              : '指定が必要です'}
          </span>
        </div>

        {article.published !== undefined && (
          <div className="content-header__row">
            <span className="content-header__row-title">published</span>
            <span className="content-header__row-result">
              {article.published ? '✅（公開）' : 'false（下書き）'}
            </span>
          </div>
        )}

        <div className="content-header__row">
          <span className="content-header__row-title">type</span>
          <span className="content-header__row-result">
            {article.type || '指定が必要です'}
            {article.type === 'tech' ? '（技術記事）' : ''}
            {article.type === 'idea' ? '（アイデア）' : ''}
          </span>
        </div>

        {!!errorCount && (
          <div>
            <div className="content-header__error-title">
              {errorCount}件の修正が必要です
            </div>
            {errorMessages.map((errorMessage, index) => (
              <ErrorRow errorMessage={errorMessage} key={`invldmsg${index}`} />
            ))}
          </div>
        )}

        <a
          href="https://zenn.dev/zenn/articles/zenn-cli-guide"
          className="content-header__link"
          rel="noopener noreferrer"
          target="_blank"
        >
          記事の作成方法 →
        </a>
      </ContentWrapper>
    </header>
  );
};
