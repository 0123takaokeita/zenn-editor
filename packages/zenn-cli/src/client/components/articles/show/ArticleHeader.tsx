import React, { useMemo } from 'react';
import styled from 'styled-components';
import { Article } from '../../../../common/types';
import { getArticleErrors } from '../../../lib/validator';
import { ContentContainer } from '../../ContentContainer';
import { TopicList } from '../../TopicList';
import { PropertyRow } from '../../PropertyRow';
import { ValidationErrors } from '../../ValidationErrors';

type Props = { article: Article };

export const ArticleHeader: React.VFC<Props> = ({ article }) => {
  const validationErrors = useMemo(
    () => getArticleErrors(article).concat(article.errors || []),
    [article]
  );

  return (
    <StyledArticleHeader>
      <ContentContainer>
        {typeof article.emoji === 'string' && (
          <div className="article-header__emoji">{article.emoji}</div>
        )}

        <h1 className="article-header__title">
          {article.title || 'titleを指定してください'}
        </h1>
        <div className="article-header__properties">
          <PropertyRow title="slug">{article.slug}</PropertyRow>

          <PropertyRow title="published">
            {typeof article.published === 'boolean' ? (
              <>{article.published ? 'true（公開）' : 'false（下書き）'}</>
            ) : (
              'true もしくは false を指定してください'
            )}
          </PropertyRow>

          <PropertyRow title="type">
            {article.type === 'tech' ? (
              'tech（技術記事）'
            ) : (
              <>
                {article.type === 'idea'
                  ? 'idea（アイデア）'
                  : 'tech もしくは idea を指定してください'}
              </>
            )}
          </PropertyRow>

          <PropertyRow title="topics">
            {Array.isArray(article.topics) && article.topics.length ? (
              <TopicList topics={article.topics} />
            ) : (
              '指定が必要です'
            )}
          </PropertyRow>
        </div>
        {!!validationErrors.length && (
          <div className="article-header__validation-errors">
            <ValidationErrors validationErrors={validationErrors} />
          </div>
        )}
      </ContentContainer>
    </StyledArticleHeader>
  );
};

const StyledArticleHeader = styled.header`
  padding: 3.3rem 0;
  background: var(--c-gray-bg);
  .article-header__emoji {
    margin-top: -0.5em;
    font-size: 56px;
    color: #000;
  }
  .article-header__title {
    font-size: 1.9rem;
  }
  .article-header__properties {
    margin-top: 1rem;
  }
  .article-header__validation-errors {
    margin-top: 1.4rem;
  }
`;
