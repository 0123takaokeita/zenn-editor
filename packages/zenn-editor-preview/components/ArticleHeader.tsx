import { Article } from "@types";
import { getArticleErrors } from "@utils/validator";
import ContentWrapper from "@components/ContentWrapper";
import ErrorRow from "@components/ErrorRow";

const ArticleHeader: React.FC<{ article: Article }> = ({ article }) => {
  const errorMessages = getArticleErrors(article);
  const errorCount = errorMessages?.length;
  return (
    <header className="content-header">
      <ContentWrapper>
        <h1 className="content-header__title">{article.title || "No Title"}</h1>

        <div className="content-header__row">
          <span className="content-header__row-title">emoji</span>

          {article.emoji ? (
            <span style={{ fontSize: "20px" }}>{article.emoji}</span>
          ) : (
            "指定が必要です"
          )}
        </div>

        <div className="content-header__row">
          <span className="content-header__row-title">slug</span>
          {article.slug}
        </div>

        <div className="content-header__row">
          <span className="content-header__row-title">topics</span>
          {article.topics
            ? article.topics.map((t) => (
                <span className="content-header__topic">{t}</span>
              ))
            : "指定が必要です"}
        </div>

        {article.public !== undefined && (
          <div className="content-header__row">
            <span className="content-header__row-title">public</span>
            {article.public?.toString()}
            {article.public === false ? "（非公開）" : ""}
          </div>
        )}

        <div className="content-header__row">
          <span className="content-header__row-title">type</span>
          {article.type || "指定が必要です"}
          {article.type === "tech" ? "（技術記事）" : ""}
          {article.type === "idea" ? "（アイデア）" : ""}
        </div>

        {!!errorCount && (
          <div>
            <div className="content-header__error-title">
              {errorCount}件のエラー
            </div>
            {errorMessages.map((errorMessage, index) => (
              <ErrorRow errorMessage={errorMessage} key={`invldmsg${index}`} />
            ))}
          </div>
        )}

        <a href="todo" className="content-header__link" target="_blank">
          Articleのmdファイルの作成方法はこちら →
        </a>
      </ContentWrapper>
    </header>
  );
};
export default ArticleHeader;
