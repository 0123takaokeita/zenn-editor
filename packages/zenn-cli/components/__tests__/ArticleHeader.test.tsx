import React from "react";
import { ArticleHeader } from "@components/ArticleHeader";
import { Article } from "@types";
import { createRenderer } from "react-test-renderer/shallow";

const renderer = createRenderer();
const dummyArticle: Article = {
  content: "test",
  emoji: "h",
  published: false,
  slug: "dummy-article",
  tags: [],
  title: "test article",
  topics: ["Zenn"],
  type: "idea",
};
describe("<ArticleHeader />", () => {
  it("should render and match the snapshot", () => {
    renderer.render(<ArticleHeader article={dummyArticle} />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
