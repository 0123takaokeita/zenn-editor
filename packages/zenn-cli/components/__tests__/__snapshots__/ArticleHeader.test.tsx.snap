// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleHeader /> should render and match the snapshot 1`] = `
<header
  className="content-header"
>
  <Unknown>
    <h1
      className="content-header__title"
    >
      test article
    </h1>
    <div
      className="content-header__row"
    >
      <span
        className="content-header__row-title"
      >
        slug
      </span>
      <span
        className="content-header__row-result"
      >
        dummy-article
      </span>
    </div>
    <div
      className="content-header__row"
    >
      <span
        className="content-header__row-title"
      >
        emoji
      </span>
      <span
        className="content-header__row-result"
      >
        <span
          style={
            Object {
              "fontSize": "20px",
            }
          }
        >
          👩‍💻
        </span>
      </span>
    </div>
    <div
      className="content-header__row"
    >
      <span
        className="content-header__row-title"
      >
        topics
      </span>
      <span
        className="content-header__row-result"
      >
        <span
          className="content-header__topic"
        >
          Zenn
        </span>
      </span>
    </div>
    <div
      className="content-header__row"
    >
      <span
        className="content-header__row-title"
      >
        published
      </span>
      <span
        className="content-header__row-result"
      >
        false（下書き）
      </span>
    </div>
    <div
      className="content-header__row"
    >
      <span
        className="content-header__row-title"
      >
        type
      </span>
      <span
        className="content-header__row-result"
      >
        idea
        
        （アイデア）
      </span>
    </div>
    <a
      className="content-header__link"
      href="https://zenn.dev/zenn/articles/zenn-cli-guide"
      rel="noopener noreferrer"
      target="_blank"
    >
      記事の作成方法 →
    </a>
  </Unknown>
</header>
`;
